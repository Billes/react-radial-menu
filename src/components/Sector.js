import React, { Component } from 'react'
import { getCircleX, getCircleY, getX, getY, getAngleBetween } from '../utils/calculations'

class Sector extends Component{

  getSectorData(r, B, C, centerX = r, centerY = r){
    const x1 = getCircleX(r, B)
    const y1 = getCircleY(r, B)
    const x2 = getCircleX(r, C)
    const y2 = getCircleY(r, C)

    return `M${centerX},${centerY} L${x1},${y1} A${r},${r} 0 0 1 ${x2},${y2} z`
  }

  renderSector({ isHovered, parameters: { radius, B, C } }){
    const fill = isHovered ? '#30C0FF' : '#E0E0E0'

    return (
      <path
        style={ { transition: 'fill 150ms ease' } }
        d={ this.getSectorData(radius, B, C) }
        fill={ fill }
        stroke='#CECECE'
        strokeWidth='1'
      />
    )
  }

  renderImage({ item, parameters: { radius, B, C } }){
    return (
      <image
        style={ { color: '#404040' } }
        transform={ `translate(-${ radius * 0.1 } -${ radius * 0.1 })` }
        width={ `${radius * 0.2}px` }
        height={ `${radius * 0.2}px` }
        x={ getX(radius, (radius * 0.78), getAngleBetween(B, C)) }
        y={ getY(radius, (radius * 0.78), getAngleBetween(B, C)) }
        xlinkHref={ item.image }
      />
    )
  }

  render(){
    const { key, onHoverChange, item } = this.props

    return (
      <a
        onMouseEnter={ () => onHoverChange(key) }
        onMouseLeave={ () => onHoverChange(key) }
        href={ item.action }
      >
        { this.renderSector(this.props) }
        { this.renderImage(this.props) }
      </a>
    )
  }
}

Sector.propTypes = {

  isHovered: React.PropTypes.bool,
  item: React.PropTypes.shape({
    action: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.func,
    ]),
    image: React.PropTypes.string,
    name: React.PropTypes.string,
  }),
  key: React.PropTypes.number,
  onHoverChange: React.PropTypes.func,
  parameters: React.PropTypes.shape({
    B: React.PropTypes.number,
    C: React.PropTypes.number,
    radius: React.PropTypes.number,
  })
}

export default Sector
