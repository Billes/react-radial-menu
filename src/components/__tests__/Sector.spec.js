/*global beforeEach, expect, jest, test */
import React from 'react'
import renderer from 'react-test-renderer'
import Sector from '../Sector'

let props

beforeEach(() => {
  props = {
    isHovered: false,
    item: { action: 'http://google.se', image: '/path/to/img', name: 'whoops' },
    onHoverChange: () => 33,
    parameters: {
      radius: 200,
      B: -90,
      C: -50,
    }
  }
})

const component = (props) => renderer.create(
    <Sector {...props} />
)

test('When hovered over sector, onHoverChange should be called', () => {
  const mock = jest.fn()
  props.onHoverChange = mock

  let tree = component(props).toJSON()
  tree.props.onMouseEnter()

  expect(mock).toHaveBeenCalled()
})

test('Fill color of sector changes when prop isHovered toggles', () => {
  props.isHovered = false
  let tree = component(props).toJSON()

  expect(tree).toMatchSnapshot()

  props.isHovered = true
  tree = component(props).toJSON()

  expect(tree).toMatchSnapshot()
})
