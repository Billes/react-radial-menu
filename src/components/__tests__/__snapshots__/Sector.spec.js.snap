// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fill color of sector changes when prop isHovered toggles 1`] = `
<a
  href="http://google.se"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  target="_blank"
>
  <path
    d="M200,200 L200,0 A200,200 0 0 1 328.5575219373079,46.791111376204384 z"
    fill="#E0E0E0"
    stroke="#CECECE"
    strokeWidth="1"
    style={
      Object {
        "transition": "fill 150ms ease",
      }
    }
  />
  <image
    height="40px"
    style={
      Object {
        "color": "#404040",
      }
    }
    transform="translate(-20 -20)"
    width="40px"
    x={253.35514235880433}
    xlinkHref="/path/to/img"
    y={53.407951157398315}
  />
</a>
`;

exports[`Fill color of sector changes when prop isHovered toggles 2`] = `
<a
  href="http://google.se"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  target="_blank"
>
  <path
    d="M200,200 L200,0 A200,200 0 0 1 328.5575219373079,46.791111376204384 z"
    fill="#30C0FF"
    stroke="#CECECE"
    strokeWidth="1"
    style={
      Object {
        "transition": "fill 150ms ease",
      }
    }
  />
  <image
    height="40px"
    style={
      Object {
        "color": "#404040",
      }
    }
    transform="translate(-20 -20)"
    width="40px"
    x={253.35514235880433}
    xlinkHref="/path/to/img"
    y={53.407951157398315}
  />
</a>
`;

exports[`render when action is a fn 1`] = `
<a
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <path
    d="M200,200 L200,0 A200,200 0 0 1 328.5575219373079,46.791111376204384 z"
    fill="#E0E0E0"
    stroke="#CECECE"
    strokeWidth="1"
    style={
      Object {
        "transition": "fill 150ms ease",
      }
    }
  />
  <image
    height="40px"
    style={
      Object {
        "color": "#404040",
      }
    }
    transform="translate(-20 -20)"
    width="40px"
    x={253.35514235880433}
    xlinkHref="/path/to/img"
    y={53.407951157398315}
  />
</a>
`;

exports[`render when action is a link 1`] = `
<a
  href="http://google.se"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  target="_blank"
>
  <path
    d="M200,200 L200,0 A200,200 0 0 1 328.5575219373079,46.791111376204384 z"
    fill="#E0E0E0"
    stroke="#CECECE"
    strokeWidth="1"
    style={
      Object {
        "transition": "fill 150ms ease",
      }
    }
  />
  <image
    height="40px"
    style={
      Object {
        "color": "#404040",
      }
    }
    transform="translate(-20 -20)"
    width="40px"
    x={253.35514235880433}
    xlinkHref="/path/to/img"
    y={53.407951157398315}
  />
</a>
`;
